File: Transport.cs
using HidApi;

namespace GenelecApp
{
    internal class Transport : IDisposable
    {
        public Transport(DeviceInfo deviceInfo)
        {
            HidDevice = deviceInfo.ConnectToDevice();
        }

        public void Send(GNetMessage message)
        {
            var data = Escape(message.Message);

            var payload = new List<byte>();
            payload.Add(0x0);
            payload.Add((byte)(0x80 + data.Count));
            payload.AddRange(data);
#if DEBUG
            Console.WriteLine($"Send command: {BitConverter.ToString(payload.ToArray())}");
#endif
            payload.AddRange(Enumerable.Repeat((byte)0x0, 65 - payload.Count));

            Console.WriteLine($"Send command: {BitConverter.ToString(payload.AsReadOnly().ToArray())}");
            

            Thread.Sleep(3);
            try
            {
                HidDevice.Write(payload.AsReadOnly().ToArray());
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Write error occurred: {ex}");
            }
            
        }

        private static List<byte> Escape(List<byte> msg)
        {
            // Perform PPP stuffing
            if (msg[^1] != Constants.GNET_TERM)
            {
                throw new ArgumentException("The last byte in the message must be GNET_TERM.");
            }

            // The message up to the terminator should be escaped with "PPP stuffing"
            List<byte> escapedMsg = new List<byte>();

            for (int i = 0; i < msg.Count - 1; i++)
            {
                byte currentByte = msg[i];

                if (currentByte == 0x7D)
                {
                    escapedMsg.Add(0x7D);
                    escapedMsg.Add(0x5D);
                }
                else if (currentByte == 0x7E)
                {
                    escapedMsg.Add(0x7D);
                    escapedMsg.Add(0x5E);
                }
                else
                {
                    escapedMsg.Add(currentByte);
                }
            }

            escapedMsg.Add(Constants.GNET_TERM);

            return escapedMsg;
        }

        public void Dispose()
        {
            HidDevice.Dispose();
        }

        public Device HidDevice { get; }
    }
}

File: src/transport.rs
use crate::{constants::GNET_TERM, gnet_message::GNetMessage};
use hidapi::{HidDevice, HidError};
use std::io;
use std::{thread::sleep, time::Duration};

pub struct Transport {
    device: HidDevice,
}

impl Transport {
    pub fn new(device: HidDevice) -> Self {
        Transport { device }
    }

    pub fn send(&self, msg: &GNetMessage) -> Result<(), HidError> {
        let data = escape(&msg.message).map_err(|e| HidError::IoError {
            error: io::Error::new(io::ErrorKind::InvalidData, e),
        })?;
        let mut payload = Vec::with_capacity(65);
        payload.push(0x00);
        payload.push(0x80u8.wrapping_add(data.len() as u8));
        payload.extend_from_slice(&data);
        payload.extend(std::iter::repeat_n(0x00, 65 - payload.len()));

        sleep(Duration::from_millis(150));

        self.device.write(&payload)?;
        Ok(())
    }
}

fn escape(msg: &[u8]) -> Result<Vec<u8>, String> {
    if *msg.last().unwrap() != GNET_TERM {
        return Err("Last byte must be GNET_TERM (0x7E)".into());
    }
    let mut escaped = Vec::new();
    for &b in &msg[..msg.len() - 1] {
        match b {
            0x7D => {
                escaped.push(0x7D);
                escaped.push(0x5D);
            }
            0x7E => {
                escaped.push(0x7D);
                escaped.push(0x5E);
            }
            other => escaped.push(other),
        }
    }
    escaped.push(GNET_TERM);
    Ok(escaped)
}
